pwiz_tools/Skyline/Util/Extensions/UtilDB.cs

```diff
@@ -119,9 +163,46 @@ public static byte[] Compress(this byte[] uncompressed, int level)
             byte[] result;
             using (var ms = new MemoryStream())
             {
-                using (var compressor =
-                        new ZlibStream(ms, CompressionMode.Compress, CompressionLevel.Level0+level))
-                    compressor.Write(uncompressed, 0, uncompressed.Length);
+                var Z = GetZlibCodec(level);
+                Z.InputBuffer = uncompressed;
+                Z.NextIn = 0;
+                Z.AvailableBytesIn = uncompressed.Length;
+                Z.OutputBuffer = WorkingBuffer;
+
+                do
+                {
+                    Z.NextOut = 0;
+                    Z.AvailableBytesOut = _workingBuffer.Length;
+                    switch (Z.Deflate(FlushType.None))
+                    {
+                        case 0:
+                        case 1:
+                            ms.Write(_workingBuffer, 0, _workingBuffer.Length - Z.AvailableBytesOut);
+                            continue;
+                        default:
+                            throw new ZlibException(@"deflating: " + Z.Message);
+                    }
+                }
+                while (Z.AvailableBytesIn > 0 || Z.AvailableBytesOut == 0);
+
+                do
+                {
+                    Z.NextOut = 0;
+                    Z.AvailableBytesOut = _workingBuffer.Length;
+                    switch (Z.Deflate(FlushType.Finish))
+                    {
+                        case 0:
+                        case 1:
+                            if (_workingBuffer.Length - Z.AvailableBytesOut > 0)
+                                ms.Write(_workingBuffer, 0, _workingBuffer.Length - Z.AvailableBytesOut);
+                            continue;
+                        default:
+                            throw new ZlibException(@"deflating: " + Z.Message);
+                    }
+                }
+                while (Z.AvailableBytesIn > 0 || Z.AvailableBytesOut == 0);
+                Z.ResetDeflate();
+

```

Comment:
@nickshulman @brendanx67 A review here would be appreciated. This code is used not just to compress the BlibDb's m/z and intensity arrays, but also the scan id arrays and point arrays in ChromCacheBuilder. It was the latter usage that required me to make the ZlibCodecs thread-static so they were parallel-safe. I haven't added a way to deallocate these static objects/buffers, but since there's only 3 of them, is it worth worrying about? I wasn't able to find a definitive answer for how you're supposed to ensure that thread-static stuff is destroyed when the thread is destroyed, especially when the thread pool is involved.

Reply:
The reason for this optimization was excessive time spent initializing zlib buffers (every time an individual m/z or intensity array was compressed for writing a blib), identified by profiling. I discussed it with Nick when I was working on it, but that was before I knew I had to make it parallel-safe because ChromCacheBuilder was using it in parallel.

This change gives a big boost to Compress performance, but I forgot to record timings. The key thing is that profiling shows that the vast majority of time spent in writing blib files is now spent in SQLite insertion I/O, not in NHibernate or zlib. This might even be an appreciable speed-up for the other code using Compress, but I haven't profiled/measured that.

Reply:
Instead of making it thread-static, can you just make it so the caller can pass the ZLibCodec into the Compress method?

ChromCacheBuilder is thread-safe when it is compressing bytes. One ChromCacheBuilder gets created for each raw file being imported. The ChromCacheBuilder (or actually, its base class "ChromCacheWriter") should create the ZLibCodec at the beginning of the import operation and get rid of it at the end.

(Also, ChromCacheMinimizer should create its own codec)

There are other places in the code where we compress things, but for most of those we don't really care about performance so it's fine if those places have to create a new ZLibCodec for each compression operation.

Reply:
How would that work with calling it from a ParallelEx loop?

Reply:
Good point. Is there a place where we do that? 

We should probably write a variation of ParallelEx.ForEach where you can specify a function to be called per thread, and that function's return value gets passed to the body action.

Reply:
Couldn't you just use a lambda expression inside the caller that passes in the value?